rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,501 @@
+- (-) A variant that references the previous node in the AST should be avoided to reduce complexity.
+- (+) Relation between documentation and the referenced implementation is straightforward and back-trackable.
+- (-) Concrete Implementation might be complex.
+- (-) Unclear if complete documentation might also need backward references. 
+    - (-), e.g., rust uses backward references only at the top of file comments.
+    - (+) Nix files can only have ONE expression; the next AST Node, in case of top-of-file comments, is thus always only that one expression. (Unlike in Rust)
+- (-) Tools need to be smart enough to understand asignments `=` and other forms of creating names for anonymous expressions. (e.g., `callPackage` and `import` )
+    - (+) Tools can still come up with other solutions that do not involve calculating everything dynamically from nix code but could also involve a static configuration.
+    - (+) The whole `tool point` is an implementation detail as long as it is not impossible. The current tool, `nixdoc`, already proves that it is possible to have static documentation to a certain degree.
+
+</details>
+
+## Single-line doc-comments (do not exist)
+
+**Observing**: Nix offers two variants of comments; single- (`#`) and multi-line comments (`/* */`). There may be use cases where it is desirable to have a form of single-line comments subtyped for doc-comment purposes. 
+
+**Considering**: Single-line comment for documentation.
+
+**Decision**: Single-line comments **cannot be used** in any form for documentation puposes.
+
+<details>
+<summary>Arguments</summary>
+
+- (-) Doc-comments should have only one variant to reduce complexity.  
+- (-) documentation will likely take up more than one line.
+- (-) If documentation grows bigger than one line, refactoring into a multiline-doc-comment must occur.
+- (+) It Would be consistent with providing variants for both nix comments.
+- (+) Offer the choice.
+- (o) Single lines can also be concatenated to form multi-line documentation.
+- (+) Takes up less vertical space
+- (-) Visually confusing when every line starts with a reserved character.
+    - (-) Potential visual conflicts with the content
+- (+) Indentation of the content is clear.
+
+</details>
+
+## Doc-comment examples
+
+The following examples demonstrate the concrete usage scenarios:
+
+`somefile.nix`
+
+~~~nix
+{
+  /**
+  Documentation for mapAttrs
+  
+  # Example
+  
+  ```
+  mapAttrs {attr = "foo"; } ...
+  ```
+  
+  # Type
+  
+  ```
+  mapAttrs :: a -> b -> c
+  ```
+  
+  */
+  mapAttrs = f: s: <...>;
+}
+~~~
+
+## Examples
+
+This section contains examples for the different use cases we would end up with; Visualize them and emphasize the previously discussed characteristics.
+
+### Attribute bindings 
+
+`somefile.nix`
+
+~~~nix
+{
+  /**
+  mapAttrs is a well-known function 
+  
+  # Examples
+  
+  "`Nix
+  some code examples
```

Comment:
> Besides that, it would be nice to have an optional output section

I guess you mean to have something like `->` or `>>>` 
to indicate what an example's evaluation should yield. So that we could unit-test our examples.

tbh. I don't think this should be targeted within this RFC (keep the scope small and clean; And address only one thing) 

But you could write tooling for that and base your own RFC on this one.

Reply:
Yeah. I agree.  
I just thought that if we provide code examples this RFC do not cover, we could have both with omitted output and with output present as some functions in `/lib` do have it as well.   
Even without RFC coverage, example would help to have less variations as it likely will be followed in same way as currently it is practice of modeling comments after lib.attrsets. And migration of current comments in `/lib` to new format would touch on this topic anyway.

Reply:
What format would you suggest?

Hint: it should be tool parseable, it cannot be -> as this is already a token from nix (Token_Implication). 

Then we can provide some examples without specifying the format

Reply:
As it is not subject of this RFC - it can be any. But it is exactly as you said. It should be parseable in order for automating changes to whatever we decide later.  
If we are using code blocks, then it might look like
````markdown
# Examples
```nix
2 + 3
```
```output
5
```
```nix
"hi\nthere\n"
```
```output-raw
hi
there
```
````
or to save space maybe
````markdown
# Examples
```nix
2 + 3
# => 5
```
```nix
"hi\nthere\n"
# raw => hi
# there
```
````

Reply:
How would you test:
Function call the needs other stuff in their scope? Like e.g lib.mapAttrs‘ 

If i used some example code i‘d use only one single md code fence maybe with some special annotations

And seperate the output from the input with => 
Then i‘d run something like proptest (its a rust crate) to check if the output matches

Reply:
I'm familiar with Haddock a bit. I think all examples acts as a REPL session and you can re-use any bindings from previous example. There is also `$setup` section outside of particular documentation if you need to keep something hidden and shared across multiple exported symbols. E.g.

````markdown
```nix
with import ../.;
lib.attrset.attrNames { x = 41; }
# => [ "x" ]
```
````

Anyway. Solving problems of doctest is a different RFC as I understand 😛

Reason why I want example is so more people would model their tests after it and it will be easier to migrate to different format with help of tooling.  
Since no one will tailor those examples for doctesting right now - I wouldn't expect them to work even if there will be right format described, but no tooling to validate it.